pkg/vcdclient/gateway.go

```diff
@@ -1049,3 +1061,109 @@ func (client *Client) IsNSXTBackedGateway() bool {
 
 	return isNSXTBackedGateway
 }
+
+func (client *Client) GetRDEVirtualIps(ctx context.Context) ([]string, string, *swaggerClient.DefinedEntity, error) {
+	defEnt, _, etag, err := client.apiClient.DefinedEntityApi.GetDefinedEntity(ctx, client.ClusterID)
+	if err != nil {
+		return nil, "", nil, fmt.Errorf("error when getting defined entity: [%v]", err)
+	}
+	statusEntry, ok := defEnt.Entity["status"]
+	if !ok {
+		return nil, "", nil, fmt.Errorf("could not find 'status' entry in defined entity")
+	}
+	statusMap, ok := statusEntry.(map[string]interface{})
+	if !ok {
+		return nil, "", nil, fmt.Errorf("unable to convert [%T] to map", statusEntry)
+	}
+	virtualIpInterfaces := statusMap["virtual_IPs"]
+	if virtualIpInterfaces == nil {
+		return make([]string, 0), etag, &defEnt, nil
+	}
+
+	virtualIpInterfacesSlice, ok := virtualIpInterfaces.([]interface{})
+	if !ok {
+		return nil, "", nil, fmt.Errorf("unable to convert [%T] to slice of interface", virtualIpInterfaces)
+	}
+	virtualIpStrs := make([]string, len(virtualIpInterfacesSlice))
+	for ind, ipInterface := range virtualIpInterfacesSlice {
+		currIp, ok := ipInterface.(string)
+		if !ok {
+			return nil, "", nil, fmt.Errorf("unable to convert [%T] to string", ipInterface)
+		}
+		virtualIpStrs[ind] = currIp
+	}
+	return virtualIpStrs, etag, &defEnt, nil
+}
+
+// This function will modify the passed in defEnt
+func (client *Client) updateRDEVirtualIps(ctx context.Context, updatedIps []string, etag string,
+	defEnt *swaggerClient.DefinedEntity) (*http.Response, error) {
+	statusEntry, ok := defEnt.Entity["status"]
+	if !ok {
+		return nil, fmt.Errorf("could not find 'status' entry in defined entity")
+	}
+	statusMap, ok := statusEntry.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("unable to convert [%T] to map", statusEntry)
+	}
+
+	statusMap["virtual_IPs"] = updatedIps
+	_, httpResponse, err := client.apiClient.DefinedEntityApi.UpdateDefinedEntity(ctx, *defEnt, etag, client.ClusterID)
+	if err != nil {
+		return httpResponse, fmt.Errorf("error when updating defined entity: [%v]", err)
+	}
+	return nil, nil
+}
+
+func (client *Client) addVirtualIpToRDE(ctx context.Context, addIp string) error {
+	for {
+		currIps, etag, defEnt, err := client.GetRDEVirtualIps(ctx)
+		if err != nil {
+			return fmt.Errorf("error for getting current vips: [%v]", err)
+		}
+		updatedIps := append(currIps, addIp)
```

Comment:
I think we should do the same as delete and update if the value doesn't exist. Since k8s is a reconciler method, it will keep retrying and we don't want to get into an error state and add multiple values.